File: wpilibj/src/main/java/edu/wpi/first/wpilibj/simulation/DoubleSolenoidSim.java
package edu.wpi.first.wpilibj.simulation;

import edu.wpi.first.wpilibj.DoubleSolenoid;
import edu.wpi.first.wpilibj.PneumaticsBase;
import edu.wpi.first.wpilibj.PneumaticsModuleType;

/** Class to control a simulated {@link edu.wpi.first.wpilibj.DoubleSolenoid}. */
public class DoubleSolenoidSim {
  private final PneumaticsBaseSim m_module;
  private final int m_fwd;
  private final int m_rev;

  /**
   * Constructs for a solenoid on the given pneumatics module.
   *
   * @param moduleSim the PCM the solenoid is connected to.
   * @param fwd the forward solenoid channel.
   * @param rev the reverse solenoid channel.
   */
  public DoubleSolenoidSim(PneumaticsBaseSim moduleSim, int fwd, int rev) {
    m_module = moduleSim;
    m_fwd = fwd;
    m_rev = rev;
  }

  /**
   * Constructs for a solenoid on a pneumatics module of the given type and ID.
   *
   * @param module the CAN ID of the pneumatics module the solenoid is connected to.
   * @param moduleType the module type (PH or PCM)
   * @param fwd the forward solenoid channel.
   * @param rev the reverse solenoid channel.
   */
  public DoubleSolenoidSim(int module, PneumaticsModuleType moduleType, int fwd, int rev) {
    this(PneumaticsBaseSim.getForType(module, moduleType), fwd, rev);
  }

  /**
   * Constructs for a solenoid on a pneumatics module of the given type and default ID.
   *
   * @param moduleType the module type (PH or PCM)
   * @param fwd the forward solenoid channel.
   * @param rev the reverse solenoid channel.
   */
  public DoubleSolenoidSim(PneumaticsModuleType moduleType, int fwd, int rev) {
    this(PneumaticsBase.getDefaultForType(moduleType), moduleType, fwd, rev);
  }

  /**
   * Check the value of the double solenoid output.
   *
   * @return the output value of the double solenoid.
   */
  public DoubleSolenoid.Value get() {
    boolean fwdState = m_module.getSolenoidOutput(m_fwd);
    boolean revState = m_module.getSolenoidOutput(m_rev);
    if (!fwdState && !revState) {
      return DoubleSolenoid.Value.kOff;
    } else if (fwdState && !revState) {
      return DoubleSolenoid.Value.kForward;
    } else if (!fwdState && revState) {
      return DoubleSolenoid.Value.kReverse;
    } else {
      throw new AssertionError(
          "In a double solenoid, both fwd and rev can't be on at the same time.");
    }
  }

  /**
   * Set the value of the double solenoid output.
   *
   * @param value The value to set (Off, Forward, Reverse)
   */
  public void set(final DoubleSolenoid.Value value) {
    boolean forward;
    boolean reverse;

    switch (value) {
      case kOff:
        forward = false;
        reverse = false;
        break;
      case kForward:
        forward = true;
        reverse = false;
        break;
      case kReverse:
        forward = false;
        reverse = true;
        break;
      default:
        throw new AssertionError("Illegal value: " + value);
    }

    m_module.setSolenoidOutput(m_fwd, forward);
    m_module.setSolenoidOutput(m_rev, reverse);
  }

  /**
   * Get the wrapped {@link PneumaticsBaseSim} object.
   *
   * @return the wrapped {@link PneumaticsBaseSim} object.
   */
  public PneumaticsBaseSim getPCMSim() {
    return m_module;
  }
}

File: wpilibc/src/main/native/cpp/simulation/DoubleSolenoidSim.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/simulation/DoubleSolenoidSim.h"

#include "frc/PneumaticsBase.h"

using namespace frc;
using namespace frc::sim;

DoubleSolenoidSim::DoubleSolenoidSim(
    std::shared_ptr<PneumaticsBaseSim> moduleSim, int fwd, int rev)
    : m_module{std::move(moduleSim)}, m_fwd{fwd}, m_rev{rev} {}

DoubleSolenoidSim::DoubleSolenoidSim(int module, PneumaticsModuleType type,
                                     int fwd, int rev)
    : m_module{PneumaticsBaseSim::GetForType(module, type)},
      m_fwd{fwd},
      m_rev{rev} {}

DoubleSolenoidSim::DoubleSolenoidSim(PneumaticsModuleType type, int fwd,
                                     int rev)
    : m_module{PneumaticsBaseSim::GetForType(
          PneumaticsBase::GetDefaultForType(type), type)},
      m_fwd{fwd},
      m_rev{rev} {}

DoubleSolenoid::Value DoubleSolenoidSim::Get() const {
  bool fwdState = m_module->GetSolenoidOutput(m_fwd);
  bool revState = m_module->GetSolenoidOutput(m_rev);
  if (fwdState && !revState) {
    return DoubleSolenoid::Value::kForward;
  } else if (!fwdState && revState) {
    return DoubleSolenoid::Value::kReverse;
  } else {
    return DoubleSolenoid::Value::kOff;
  }
}

void DoubleSolenoidSim::Set(DoubleSolenoid::Value output) {
  m_module->SetSolenoidOutput(m_fwd, output == DoubleSolenoid::Value::kForward);
  m_module->SetSolenoidOutput(m_rev, output == DoubleSolenoid::Value::kReverse);
}

bool DoubleSolenoid::IsForward() const {
  return Get() == kForward;
}

bool DoubleSolenoid::IsReverse() const {
  return Get() == kReverse;
}

bool DoubleSolenoid::IsOff() const {
  return Get() == kOff;
}

void DoubleSolenoid::Forward() {
  Set(kForward);
}

void DoubleSolenoid::Reverse() {
  Set(kReverse);
}

void DoubleSolenoid::Off() {
  Set(kOff);
}

std::shared_ptr<PneumaticsBaseSim> DoubleSolenoidSim::GetModuleSim() const {
  return m_module;
}
